sdk/opa_test.go

```diff
@@ -0,0 +1,496 @@
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": `
+				package system
+
+				main = 7
+			`,
+		}),
+	)
+
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"discovery": {
+			"resource": "/bundles/discovery.tar.gz"
+		}
+	}`, server.URL()))
+
+	opa, err := sdk.New(ctx, sdk.Options{
+		Config: bytes.NewReader(config),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer opa.Stop(ctx)
+
+	exp := json.Number("7")
+
+	if result, err := opa.Decision(ctx, sdk.DecisionOptions{}); err != nil {
+		t.Fatal(err)
+	} else if result.Result != exp {
+		t.Fatalf("expected %v but got %v", exp, result.Result)
+	}
+
+}
+
+func TestAsync(t *testing.T) {
+
+	ctx := context.Background()
+
+	readyCh := make(chan struct{})
+
+	server := sdktest.MustNewServer(
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": "package system\nmain = 7",
+		}),
+	)
+
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/bundle.tar.gz"
+			}
+		}
+	}`, server.URL()))
+
+	opa, err := sdk.New(ctx, sdk.Options{
+		Config: bytes.NewReader(config),
+		Ready:  readyCh,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if result, err := opa.Decision(ctx, sdk.DecisionOptions{}); !sdk.IsUndefinedErr(err) {
+		t.Fatal("expected undefined error but got", result, "err:", err)
+	}
+
+	defer opa.Stop(ctx)
+	<-readyCh
+
+	exp := json.Number("7")
+
+	if result, err := opa.Decision(ctx, sdk.DecisionOptions{}); err != nil || !reflect.DeepEqual(result.Result, exp) {
+		t.Fatal("expected 7 but got", result, "err:", err)
+	}
+}
+
+func TestCancelStartup(t *testing.T) {
+
+	server := sdktest.MustNewServer()
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/doesnotexist.tar.gz"
+			}
+		}
+		}`, server.URL()))
+
+	// Server will return 404 responses because bundle does not exist. OPA should timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+	defer cancel()
+
+	_, err := sdk.New(ctx, sdk.Options{
+		Config: bytes.NewReader(config),
+	})
+	if !errors.Is(err, context.DeadlineExceeded) {
+		t.Fatalf("expected deadline exceeded error but got %v", err)
+	}
+}
+
+func TestConfigAsYAML(t *testing.T) {
+
+	server := sdktest.MustNewServer(
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": "package system\nmain = 7",
+		}),
+	)
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`services:
+  test:
+    url: %q
+bundles:
+  test:
+    resource: "/bundles/bundle.tar.gz"`, server.URL()))
+
+	ctx := context.Background()
+	_, err := sdk.New(ctx, sdk.Options{
+		Config: bytes.NewReader(config),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestConfigure(t *testing.T) {
```

Comment:
leaktest could be useful here, since multiple go routines are involved. Or not. But we could try it.

👉 https://github.com/open-policy-agent/opa/blob/82cca627647165804f8cf042cfd7f84dd13ddecc/rego/rego_wasmtarget_test.go#L202